### PR TITLE
implement Clone for LessSafeKey

### PR DESCRIPTION
**Pull request by oconnor663**
OpeningKey and SealingKey intentionally avoid implementing Copy and
Clone, because they're attached to a fixed nonce sequence that should be
unique. LessSafeKey isn't attached to a nonce sequence, though, and
making it Clone lets callers avoid repeating key setup work.

Coming from https://github.com/briansmith/ring/pull/1158#issuecomment-765050031.

### CODE REVIEW

**Comment by briansmith** [comment]
Please review PR #1185 and see if rebasing this on top of that PR will simplify this. AFAICT it will.

**Comment by oconnor663** [comment]
Yes it looks like the manual impl isn't needed on that branch. Everything is just `#[derive(Clone)]`: https://github.com/oconnor663/ring/commit/3275a8df3c1abb31ae52d96eeb7719ef95bdad80

**Comment by briansmith** [comment]
> Yes it looks like the manual impl isn't needed on that branch. Everything is just `#[derive(Clone)]`: [oconnor663@3275a8d](https://github.com/oconnor663/ring/commit/3275a8df3c1abb31ae52d96eeb7719ef95bdad80)

Thanks! I have now merged that refactoring. If you rebase this on top of main, and add a test that encrypting some arbitrary data with the original key results in the same output as encrypting the same arbitrary data with a clone of the key (with the same nonce), then I'll prioritize merging this and ensure it is included in the next release.

**Comment by briansmith** [comment]
@oconnor663 Are you still interested in this? I'm preparing a new release that could include this if you rebase it onto main.

**Comment by oconnor663** [comment]
Done! Thanks for the reminder.

File: tests/aead_tests.rs [hunk 1]
```diff
@@ -50,7 +50,7 @@ macro_rules! test_aead {
             $(
                 #[allow(non_snake_case)]
                 mod $alg { // Provide a separate namespace for each algorithm's test.
-                    use super::super::{*, test};
```
Comment (briansmith) [inline tests/aead_tests.rs:53]: I think you'll have to undo this change? I put this ugliness there because the tests wouldn't pass correctly in wasm32 targets without it.

File: tests/aead_tests.rs [hunk 2]
```diff
@@ -509,6 +509,38 @@ fn test_aead_key_debug() {
     );
 }
 
+#[test]
+#[cfg_attr(target_arch = "wasm32", wasm_bindgen_test)]
```
Comment (briansmith) [inline tests/aead_tests.rs:513]: The wasm_bindgen_test thing here shouldn't be necessary anymore because of what was done in PR #1252.

File: tests/aead_tests.rs [hunk 3]
```diff
@@ -509,6 +509,38 @@ fn test_aead_key_debug() {
...
+fn test_aead_lesssafekey_clone() {
```
Comment (briansmith) [inline tests/aead_tests.rs:514]: Please parameterize this on the `&'static aead::Algorithm` and then test it with all three AES-128-GCM, AES-256-GCM, and ChaCha20-Poly1305. Especially the AES and ChaCha20-Poly1305 code is completely different, it's good to cover all.

**Review by briansmith — CHANGES_REQUESTED** [review]

File: tests/aead_tests.rs (hunk 1 above)
Comment (oconnor663) [reply tests/aead_tests.rs:53]: Ok, I'll split it out into its own line with its own `#[allow(unused_imports)]`. But if you'd prefer to just leave the warning in place let me know.

**Comment by oconnor663** [comment]
Addressed the comments above. Let me know what you think about the `allow(unused_imports)` workaround in particular. Also if you want me to combine the three separate tests into one test with a loop.

File: tests/aead_tests.rs (hunk 1 above)
Comment (briansmith) [reply tests/aead_tests.rs:53]: I understand now that you're just trying to fix a warning. It turns out that our warning detection in CI was broken. I fixed it in PR #1260. So, you should be able to rebase this on top of the current `main` and you shouldn't need to make changes here any more.

File: tests/aead_tests.rs [hunk 4]
```diff
@@ -509,6 +512,52 @@ fn test_aead_key_debug() {
...
+    let tag2 = key2
+        .seal_in_place_separate_tag(
+            aead::Nonce::try_assume_unique_for_key(&nonce_bytes).unwrap(),
+            aead::Aad::empty(),
+            &mut buf2,
+        )
+        .unwrap();
+    assert_eq!(tag1.as_ref(), tag2.as_ref());
+    assert_eq!(buf1, buf2);
+}
```
Comment (briansmith) [inline tests/aead_tests.rs:544]: I wonder why codecov is saying this line isn't hit?

**Comment by briansmith** [comment]
@oconnor663 If you can rebase this and address the comment above, then I think we're good to go and I'll be able to include this in the next release.

**Comment by oconnor663** [comment]
I've rebased the branch and removed the extra commit that was just fixing a warning. Not sure about the codecov message. If I put a `panic!()` at the end of that function, `cargo test` fails as expected, so those lines are certainly getting run. I haven't yet figured out how to run codecov locally, but if you think that's the best next step let me know.

It looks like some repo settings have changed, and the CI run for the rebased branch won't start until I get some new permissions. Let me know if you need anything from me for that.

File: tests/aead_tests.rs (hunk 4 above)
Comment (briansmith) [reply tests/aead_tests.rs:545]: Let's now worry about the codecov thing.

**Comment by briansmith** [comment]
> It looks like some repo settings have changed, and the CI run for the rebased branch won't start until I get some new permissions. Let me know if you need anything from me for that.

I started the CI run. Recently GitHub started requiring manual approval for the first PR from a new contributor to a repo, to prevent cryptocurrency mining and other abuse of GitHub Actions.

**Comment by oconnor663** [comment]
Oof. As they say about nice things, and why we cannot have them...

**Review by briansmith — APPROVED** [review]

**Comment by briansmith** [comment]
Thanks! Merged.

**Comment by oconnor663** [comment]
Great! Thanks for all the feedback.
